src/cart/crt_group.c

```diff
@@ -3150,122 +3150,118 @@ int crt_group_primary_rank_add(crt_context_t ctx, crt_group_t *grp,
 }
 
 /*
- * Helper function to return back rank lists of nodes to add,
- * to remove, and index of ranks to be added.
- *
- * Index of ranks to add is used for accessing proper URIs
- * when adding new ranks
+ * Helper function to return back a list of indices (into mod_membs) to add and
+ * a list of ranks to remove.
  */
 static int
-crt_group_mod_get(d_rank_list_t *grp_membs, d_rank_list_t *mod_membs,
-		  crt_group_mod_op_t op, d_rank_list_t **ret_to_add,
-		  d_rank_list_t **ret_to_remove, uint32_t **ret_idx_to_add)
+crt_group_mod_get(struct crt_grp_priv *grp_priv, d_rank_list_t *mod_membs, uint64_t *incarnations,
+		  crt_group_mod_op_t op, uint32_t **ret_to_add, uint32_t *ret_n_to_add,
+		  d_rank_list_t **ret_to_remove)
 
 {
+	d_rank_list_t	*grp_membs;
 	d_rank_t	rank;
 	int		rc = 0;
-	d_rank_list_t	*to_add = NULL;
+	uint32_t	*to_add = NULL;
+	uint32_t	n_to_add;
 	d_rank_list_t	*to_remove = NULL;
-	uint32_t	*idx_to_add = NULL;
 	int		i;
 
+	grp_membs = grp_priv_get_membs(grp_priv);
+
 	D_ASSERT(grp_membs != NULL);
 	D_ASSERT(mod_membs != NULL);
 	D_ASSERT(ret_to_add != NULL);
 	D_ASSERT(ret_to_remove != NULL);
-	D_ASSERT(ret_idx_to_add != NULL);
 
 	/* At most we will remove all members from old group */
 	to_remove = d_rank_list_alloc(grp_membs->rl_nr);
 
 	/* At most we will add all members from new group */
-	to_add = d_rank_list_alloc(mod_membs->rl_nr);
+	D_ALLOC_ARRAY(to_add, mod_membs->rl_nr);
 
 	if (to_remove == NULL || to_add == NULL) {
 		D_ERROR("Failed to allocate lists\n");
 		D_GOTO(cleanup, rc = -DER_NOMEM);
 	}
 
-	/* Array will have at most 'mod_membs' elements */
-	D_ALLOC_ARRAY(idx_to_add, mod_membs->rl_nr);
-	if (!idx_to_add) {
-		D_GOTO(cleanup, rc = -DER_NOMEM);
-	}
-
-	to_add->rl_nr = 0;
+	n_to_add = 0;
 	to_remove->rl_nr = 0;
 	/* Build to_add and to_remove lists based on op specified */
 	if (op == CRT_GROUP_MOD_OP_REPLACE) {
 		/*
 		 * Replace:
 		 * If rank exists in mod_membs but not in grp_membs - add it
 		 * If rank exists in grp_membs but not in mod_membs - remove it
-		 * Otherwise leave rank unchanged
+		 * Otherwise leave rank unchanged (but may update its SWIM state)
 		 */
 
 		/* Build list of new ranks to add */
 		for (i = 0; i < mod_membs->rl_nr; i++) {
 			rank = mod_membs->rl_ranks[i];
 
-			if (d_rank_in_rank_list(grp_membs, rank) == false) {
-				idx_to_add[to_add->rl_nr] = i;
-				to_add->rl_ranks[to_add->rl_nr++] = rank;
+			if (d_rank_in_rank_list(grp_membs, rank)) {
+				if (incarnations != NULL) {
+					rc = crt_swim_rank_check(grp_priv, rank, incarnations[i]);
```

Comment:
i am not sure i like that we have a check here. i would rather have this function return: idx_to_add and idx_to_modify  and the outer logic would go through those and perform crt_swim_rank_check() on appropriate ranks that appear at those indecies